### DISCUSSION TITLE
How to apply existing static type to schema?

### DISCUSSION THREAD

**Discussion by stefan-girlich** [colinhacks/zod#3367] [Q&A]
I have a statically defined type `FormData` and I would like to apply it to my schema so I can
a) make sure that the schema matches the type
b) access the schema e.g. to check `schema.shape.name.isOptional()`
```ts
interface FormData {
    name: string
}

const schema = z.object({
    name: z.string()
});
```

I ended up extracting and typing the schema shape as following:
```ts
const schemaShape: Record<keyof FormData, ZodSchema> = {
  name: z.string(),
}
const schema = z.object(schemaShape)
```

Is there a more idiomatic way to achieve this?

**Comment by RobinTail** [comment]
I'd suggest to do the opposite — define the type from the schema:

```ts
const schema = z.object({
    name: z.string()
});

type FormData = z.infer<typeof schema>;
```

If the schema has transformations, you can use `z.input<>` or `z.output<>` depending on your needs.

, @stefan-girlich

**Comment by donnyroufs** [comment]
The previous response did not address the actual question. To provide more context: for technical reasons, I need to create an abstraction between two layers. The first layer must be entirely free of any libraries. Therefore, we need to define a contract at this layer. However, I would still like to use Zod in the other layer to validate against this contract. How can I make my Zod schema smart enough to throw me an error when it doesn't match the contract?

```ts
// Layer A
interface ICreateGithubComment {
  content: string;
}

// Layer B
const schema = z.object<ICreateGithubComment>({ content: z.string() });
```

**Reply by RobinTail** [reply]
https://www.npmjs.com/package/ts-to-zod
